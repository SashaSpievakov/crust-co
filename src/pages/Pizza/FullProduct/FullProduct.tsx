import { FC, useState } from 'react';

import { ItemsCountHandler, ProductOptions } from '@src/components';
import { ButtonLink } from '@src/components/UI';
import { IPizzaItem } from '@src/models';

import { Image, Item, Paragraph, Price, Title } from './FullProduct.styled';

interface FullProductProps {
  item: IPizzaItem;
}

export const typeNames: string[] = ['traditional', 'thin'];

export const FullProduct: FC<FullProductProps> = ({ item }) => {
  const [activePrice, setActivePrice] = useState<number>(item.price);
  const [activeSize, setActiveSize] = useState<number>(12);
  const [activeType, setActiveType] = useState<number>(0);

  return (
    <Item data-testid="itemPage">
      <picture>
        <source
          srcSet={`../assets/img/pizzas/pizza${item.id}.webp`}
          type="image/webp"
        />
        <source
          srcSet={`../assets/img/pizzas/pizza${item.id}.png`}
          type="image/png"
        />
        <Image
          src={`../assets/img/pizzas/pizza${item.id}.png`}
          alt={`${item.name} pizza`}
          width="450"
          height="450"
        />
      </picture>
      <Title>{item.name}</Title>
      <Paragraph>{item.description}</Paragraph>
      <ProductOptions
        price={activePrice}
        sizes={item.sizes}
        types={item.types}
        activeSize={activeSize}
        activeType={activeType}
        setActivePrice={setActivePrice}
        setActiveSize={setActiveSize}
        setActiveType={setActiveType}
        typeNames={typeNames}
        isFullScreen
      />
      <Price>{activePrice}$</Price>
      <ItemsCountHandler
        id={item.id}
        name={item.name}
        price={activePrice}
        activeSize={activeSize}
        activeType={activeType}
        typeNames={typeNames}
        isFullScreen
      />
      <ButtonLink link="/">Go Back</ButtonLink>
    </Item>
  );
};
