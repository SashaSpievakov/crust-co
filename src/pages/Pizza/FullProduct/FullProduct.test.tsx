import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  rendererWithProviders,
  renderWithProvidersAndRoutes,
} from '../../../tests/helpers';
import { mockItem } from '../../../tests/mocks/mockData/mockData';
import { FullProduct } from './FullProduct';

describe('FullProduct tests', () => {
  test('should render the FullProduct component', () => {
    const snapshot = rendererWithProviders(<FullProduct item={mockItem} />);
    expect(snapshot).toMatchSnapshot();
  });

  describe('FullProduct price changing', () => {
    beforeEach(() => {
      renderWithProvidersAndRoutes(<FullProduct item={mockItem} />);
    });

    test('should choose new type', () => {
      const typeItem = screen.getByText(/thin/i);
      const price = screen.getByRole('heading', {
        level: 3,
      });

      userEvent.click(typeItem);

      expect(price).toHaveTextContent('15$');
    });

    test('should choose new size', () => {
      const sizeItem = screen.getByText(/16 inch/i);
      const price = screen.getByRole('heading', {
        level: 3,
      });

      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('18$');
    });

    test('should choose new type and size', () => {
      const typeItem = screen.getByText(/thin/i);
      const sizeItem = screen.getByText(/14 inch/i);
      const price = screen.getByRole('heading', {
        level: 3,
      });

      userEvent.click(typeItem);
      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('17$');
    });
  });

  describe('FullProduct integrations', () => {
    beforeEach(() => {
      renderWithProvidersAndRoutes(<FullProduct item={mockItem} />);
    });

    test('should disable minus button after type switching', () => {
      const sizeItem = screen.getByText(/16 inch/i);
      const buttonAdd = screen.getByText(/add/i);
      userEvent.click(buttonAdd);

      const buttonPlus = screen.getByTestId('itemsHandlerPlus');
      const buttonMinus = screen.getByTestId('itemsHandlerMinus');
      const count = screen.getByText('1');

      userEvent.click(buttonPlus);
      userEvent.click(buttonPlus);

      userEvent.click(sizeItem);
      userEvent.click(buttonPlus);
      userEvent.click(buttonMinus);
      userEvent.click(buttonMinus);

      expect(count).toHaveTextContent('3');
    });

    test('should switch to add button after multiple changes in the count', () => {
      const typeTraditional = screen.getByText(/traditional/i);
      const typeThin = screen.getByText(/thin/i);
      const twelve = screen.getByText(/12 inch/i);
      const fourteen = screen.getByText(/14 inch/i);
      const sixteen = screen.getByText(/16 inch/i);
      const buttonAdd = screen.getByText(/add/i);
      userEvent.click(buttonAdd);

      const buttonPlus = screen.getByTestId('itemsHandlerPlus');
      const buttonMinus = screen.getByTestId('itemsHandlerMinus');

      userEvent.click(buttonPlus);

      userEvent.click(typeThin);
      userEvent.click(fourteen);
      userEvent.click(buttonPlus);
      userEvent.click(buttonPlus);

      userEvent.click(sixteen);
      userEvent.click(buttonPlus);

      userEvent.click(fourteen);
      userEvent.click(buttonMinus);
      userEvent.click(buttonMinus);
      userEvent.click(buttonMinus);

      userEvent.click(sixteen);
      userEvent.click(buttonMinus);

      userEvent.click(twelve);
      userEvent.click(typeTraditional);
      userEvent.click(buttonMinus);
      userEvent.click(buttonMinus);

      expect(screen.getByText(/add/i)).toBeInTheDocument();
    });
  });
});
