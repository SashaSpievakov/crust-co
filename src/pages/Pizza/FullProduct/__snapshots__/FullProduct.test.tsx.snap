// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullProduct Tests renders the FullProduct component 1`] = `
Object {
  "children": Array [
    <picture>
      <source
        srcSet="../assets/img/pizzas/pizza9.webp"
        type="image/webp"
      />
      <source
        srcSet="../assets/img/pizzas/pizza9.png"
        type="image/png"
      />
      <img
        alt="Chicken Curry pizza"
        className="sc-jTjUTQ fBjZcC"
        height="450"
        src="../assets/img/pizzas/pizza9.png"
        width="450"
      />
    </picture>,
    <h2
      className="sc-jRwbcX cxexbC"
    >
      Chicken Curry
    </h2>,
    <p
      className="sc-fvEvSO gqqqwN"
    >
      Lorem, ipsum dolor sit amet consectetur adipisicing elit. Aliquid nisi aspernatur debitis quod suscipit, vitae placeat natus ipsum inventore esse rerum animi facere numquam saepe vero mollitia quibusdam at voluptates est commodi laudantium? Iste inventore quibusdam cupiditate nemo sint iusto minus nihil culpa! Sequi architecto nesciunt explicabo mollitia. Laboriosam, odio.
    </p>,
    <div
      className="sc-hOzowv HmFZK"
    >
      <ul>
        <li
          aria-current={true}
          className="sc-dsHJmm ewokfL"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          traditional
        </li>
        <li
          aria-current={false}
          className="sc-dsHJmm dgbJWc"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          thin
        </li>
      </ul>
      <ul>
        <li
          aria-current={true}
          className="sc-dsHJmm ewokfL"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          12
           inch
        </li>
        <li
          aria-current={false}
          className="sc-dsHJmm dgbJWc"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          14
           inch
        </li>
        <li
          aria-current={false}
          className="sc-dsHJmm dgbJWc"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          16
           inch
        </li>
      </ul>
    </div>,
    <h3
      className="sc-fcCDlF bXRcff"
    >
      14
      $
    </h3>,
    <button
      className="sc-bcXHqe sc-eDvSVe sc-gKPRtg glVyWT fwWQTF dBjrMs"
      onClick={[Function]}
    >
      <svg
        className="sc-kImNAt ifacgh"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
        />
      </svg>
      <span>
        Add
      </span>
    </button>,
    <a
      className="sc-bcXHqe sc-gswNZR glVyWT bMYeao"
      href="/"
      onClick={[Function]}
    >
      Go Back
    </a>,
  ],
  "props": Object {
    "className": "sc-hhOBVt sc-cZFQFd cSuYdA epSrUx",
    "data-testid": "itemPage",
  },
  "type": "article",
}
`;
